File: Autonomous Mobile Robots/Chapter13/Lesson5/Chapter13_Lesson5.cpp
// Chapter13_Lesson5.cpp
// Minimal ATE evaluation (TUM format) + Umeyama alignment.
//
// Build (example):
//   g++ -O2 -std=c++17 Chapter13_Lesson5.cpp -I /usr/include/eigen3 -o ate_eval
//
// Run:
//   ./ate_eval groundtruth.txt estimated.txt 0.02 1
//
// Args:
//   argv[1]=gt, argv[2]=est, argv[3]=max_dt, argv[4]=allow_scale (0/1)
#include <Eigen/Dense>
#include <fstream>
#include <iostream>
#include <sstream>
#include <string>
#include <vector>
#include <unordered_set>
#include <cmath>

struct Traj {
  std::vector<double> t;
  std::vector<Eigen::Vector3d> p;
};

static Traj read_tum(const std::string& path) {
  Traj tr;
  std::ifstream f(path);
  if (!f) throw std::runtime_error("Cannot open " + path);
  std::string line;
  while (std::getline(f, line)) {
    if (line.empty() || line[0] == '#') continue;
    std::istringstream iss(line);
    double tt, tx, ty, tz, qx, qy, qz, qw;
    if (!(iss >> tt >> tx >> ty >> tz >> qx >> qy >> qz >> qw)) continue;
    tr.t.push_back(tt);
    tr.p.emplace_back(tx, ty, tz);
  }
  return tr;
}

static std::vector<std::pair<int,int>> associate_by_time(
  const std::vector<double>& ta,
  const std::vector<double>& tb,
  double max_dt)
{
  std::vector<std::pair<int,int>> pairs;
  int j = 0;
  std::unordered_set<int> used;
  for (int i = 0; i < (int)ta.size(); ++i) {
    double t = ta[i];
    while (j + 1 < (int)tb.size() && tb[j] < t) j++;
    std::vector<int> cand;
    for (int jj : {j, j-1}) {
      if (0 <= jj && jj < (int)tb.size() && used.find(jj) == used.end()) cand.push_back(jj);
    }
    if (cand.empty()) continue;
    int best = cand[0];
    for (int jj : cand) if (std::abs(tb[jj] - t) < std::abs(tb[best] - t)) best = jj;
    if (std::abs(tb[best] - t) <= max_dt) {
      pairs.emplace_back(i, best);
      used.insert(best);
    }
  }
  return pairs;
}

static void umeyama(
  const Eigen::MatrixXd& A, // 3xN
  const Eigen::MatrixXd& B, // 3xN
  bool with_scale,
  double& s,
  Eigen::Matrix3d& R,
  Eigen::Vector3d& t)
{
  const int N = (int)A.cols();
  if (N < 3) throw std::runtime_error("Need >=3 points");
  Eigen::Vector3d muA = A.rowwise().mean();
  Eigen::Vector3d muB = B.rowwise().mean();
  Eigen::MatrixXd X = A.colwise() - muA;
  Eigen::MatrixXd Y = B.colwise() - muB;

  Eigen::Matrix3d Sigma = (Y * X.transpose()) / (double)N;

  Eigen::JacobiSVD<Eigen::Matrix3d> svd(Sigma, Eigen::ComputeFullU | Eigen::ComputeFullV);
  Eigen::Matrix3d U = svd.matrixU();
  Eigen::Matrix3d V = svd.matrixV();
  Eigen::Vector3d D = svd.singularValues();

  Eigen::Matrix3d S = Eigen::Matrix3d::Identity();
  if ((U.determinant() * V.determinant()) < 0.0) S(2,2) = -1.0;

  R = U * S * V.transpose();

  if (with_scale) {
    double varA = (X.array() * X.array()).sum() / (double)N;
    s = (D.transpose() * S.diagonal()) / (varA + 1e-12);
  } else {
    s = 1.0;
  }

  t = muB - s * (R * muA);
}

int main(int argc, char** argv) {
  if (argc < 5) {
    std::cerr << "Usage: " << argv[0] << " gt.txt est.txt max_dt allow_scale(0/1)\n";
    return 2;
  }
  const std::string gt_path = argv[1];
  const std::string est_path = argv[2];
  const double max_dt = std::stod(argv[3]);
  const bool allow_scale = (std::stoi(argv[4]) != 0);

  Traj gt = read_tum(gt_path);
  Traj est = read_tum(est_path);

  auto pairs = associate_by_time(est.t, gt.t, max_dt);
  if ((int)pairs.size() < 3) {
    std::cerr << "Too few associated poses: " << pairs.size() << "\n";
    return 3;
  }

  const int N = (int)pairs.size();
  Eigen::MatrixXd A(3, N), B(3, N);
  for (int k = 0; k < N; ++k) {
    A.col(k) = est.p[pairs[k].first];
    B.col(k) = gt.p[pairs[k].second];
  }

  double s;
  Eigen::Matrix3d R;
  Eigen::Vector3d t;
  umeyama(A, B, allow_scale, s, R, t);

  Eigen::MatrixXd A_al = (s * (R * A)).colwise() + t;
  Eigen::MatrixXd E = A_al - B;
  Eigen::VectorXd se = E.colwise().squaredNorm();

  double rmse = std::sqrt(se.mean());
  double med;
  {
    std::vector<double> e;
    e.reserve(N);
    for (int k = 0; k < N; ++k) e.push_back(std::sqrt(se(k)));
    std::nth_element(e.begin(), e.begin() + e.size()/2, e.end());
    med = e[e.size()/2];
  }
  double mx = std::sqrt(se.maxCoeff());

  std::cout << "pairs " << N << "\n";
  std::cout << "scale " << s << "\n";
  std::cout << "rmse_m " << rmse << "\n";
  std::cout << "median_m " << med << "\n";
  std::cout << "max_m " << mx << "\n";
  return 0;
}

File: Autonomous_Mobile_Robots_Course/Chapter13/Lesson5/Chapter13_Lesson5.m
% Chapter13_Lesson5.m
% MATLAB utilities for VIO/Visual SLAM lab:
%  1) ATE evaluation (TUM format) with optional similarity scale.
%  2) Allan variance-based IMU noise parameter estimation (gyro/accel).
%
% Usage examples:
%   out = ate_eval_tum("groundtruth.txt","estimated.txt",0.02,true);
%   imu = readmatrix("imu.csv");  % columns: t, ax, ay, az, gx, gy, gz
%   av  = allan_imu(imu(:,1), imu(:,2:4), imu(:,5:7));
%
% NOTE: For monocular VO/SLAM, set allowScale=true.
function demo_Chapter13_Lesson5()
  disp("Run functions: ate_eval_tum, allan_imu");
end

function out = ate_eval_tum(gtPath, estPath, maxDt, allowScale)
  gt = read_tum(gtPath);
  est = read_tum(estPath);

  pairs = associate_by_time(est.t, gt.t, maxDt);
  if size(pairs,1) < 3
    error("Too few associated poses.");
  end

  A = est.p(pairs(:,1), :);
  B = gt.p(pairs(:,2), :);

  [s,R,t] = umeyama(A, B, allowScale);
  A_al = (s*(R*A') + t)';
  e = A_al - B;
  se = sum(e.^2,2);
  out.n_pairs = size(pairs,1);
  out.scale = s;
  out.R = R;
  out.t = t;
  out.rmse = sqrt(mean(se));
  out.median = median(sqrt(se));
  out.max = max(sqrt(se));
end

function tr = read_tum(path)
  fid = fopen(path,'r');
  if fid < 0, error("Cannot open file"); end
  t = [];
  p = [];
  while true
    line = fgetl(fid);
    if ~ischar(line), break; end
    line = strtrim(line);
    if isempty(line) || startsWith(line,"#"), continue; end
    parts = split(line);
    if numel(parts) < 8, continue; end
    tt = str2double(parts{1});
    tx = str2double(parts{2});
    ty = str2double(parts{3});
    tz = str2double(parts{4});
    t(end+1,1) = tt; %#ok<AGROW>
    p(end+1,:) = [tx ty tz]; %#ok<AGROW>
  end
  fclose(fid);
  tr.t = t;
  tr.p = p;
end

function pairs = associate_by_time(ta, tb, maxDt)
  pairs = [];
  j = 1;
  used = false(numel(tb),1);
  for i = 1:numel(ta)
    t = ta(i);
    while (j+1) <= numel(tb) && tb(j) < t
      j = j+1;
    end
    cand = [j, j-1];
    best = -1;
    bestErr = inf;
    for c = cand
      if c >= 1 && c <= numel(tb) && ~used(c)
        err = abs(tb(c)-t);
        if err < bestErr
          bestErr = err;
          best = c;
        end
      end
    end
    if best > 0 && bestErr <= maxDt
      pairs(end+1,:) = [i, best]; %#ok<AGROW>
      used(best) = true;
    end
  end
end

function [s,R,t] = umeyama(A, B, withScale)
  % A,B are Nx3
  n = size(A,1);
  muA = mean(A,1);
  muB = mean(B,1);
  X = A - muA;
  Y = B - muB;
  Sigma = (Y'*X)/n;

  [U,D,V] = svd(Sigma);
  S = eye(3);
  if det(U)*det(V) < 0
    S(3,3) = -1;
  end
  R = U*S*V';
  if withScale
    varA = sum(sum(X.^2))/n;
    s = trace(D*S)/(varA + 1e-12);
  else
    s = 1.0;
  end
  t = muB' - s*R*muA';
end

function av = allan_imu(t, acc, gyro)
  % Compute Allan deviation for 3-axis accel and gyro.
  % t: Nx1 seconds (monotonic). acc: Nx3 (m/s^2). gyro: Nx3 (rad/s).
  % Output av contains taus and sigma(taus) per axis.
  dt = median(diff(t));
  if ~isfinite(dt) || dt <= 0, error("Invalid dt"); end

  maxM = floor(numel(t)/10);
  Ms = unique(round(logspace(0, log10(maxM), 50)));
  taus = Ms * dt;

  av.taus = taus;
  av.acc = cell(1,3);
  av.gyro = cell(1,3);

  for ax = 1:3
    av.acc{ax} = allan_dev_1d(acc(:,ax), Ms);
    av.gyro{ax} = allan_dev_1d(gyro(:,ax), Ms);
  end

  % Rough noise parameters (white + random walk) from log-log slopes:
  % white noise ~ slope -1/2, random walk ~ slope +1/2.
  % Users typically pick two tau regions and fit manually for robust estimates.
end

function sigma = allan_dev_1d(x, Ms)
  x = x(:);
  N = numel(x);
  sigma = nan(numel(Ms),1);
  for k = 1:numel(Ms)
    m = Ms(k);
    K = floor(N/(2*m));
    if K < 2, continue; end
    % cluster averages
    y = zeros(2*K,1);
    for i = 1:(2*K)
      idx = (i-1)*m + (1:m);
      y(i) = mean(x(idx));
    end
    d = diff(y);
    sigma(k) = sqrt(0.5*mean(d.^2));
  end
end
